File: src/com/gaolei/java/MinStack.java
package com.gaolei.java;

import java.util.Stack;

class MinStack {

    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    /**
     * initialize your data structure here.
     */
    public MinStack() {
        this.stack = new Stack<>();
        this.minStack = new Stack<>();
    }

    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty()) {
            minStack.push(x);
        } else {
            Integer peek = minStack.peek();
            if (x < peek) {
                minStack.push(x);
            } else {
                minStack.push(peek);
            }
        }
    }

    public void pop() {
        stack.pop();
        minStack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}

File: src/com/gaolei/kotlin/MinStack.kt
package com.gaolei.kotlin

import java.util.*


class MinStack() {

    /** initialize your data structure here. */
    private val minStack = Stack<Int>()
    private val stack = Stack<Int>()

    fun push(x: Int) {
        stack.push(x)
        if (minStack.isEmpty()) {
            minStack.push(x)
        } else {
            val peek = minStack.peek()
            if (x < peek) minStack.push(x)
            else minStack.push(peek)
        }
    }

    fun pop() {
        stack.pop()
        minStack.pop()
    }

    fun top(): Int {
        return stack.peek()
    }

    fun getMin(): Int {
        return minStack.peek()
    }

}
